runners/flink/src/test/java/org/apache/beam/runners/flink/PortableExecutionTest.java

```diff
@@ -0,0 +1,234 @@
+  }
+
+  @Parameter public boolean isStreaming;
+
+  private transient ExecutorService sdkHarnessExecutor;
+  private transient Future<?> sdkHarnessExecutorFuture;
+
+  private DockerJobBundleFactory createJobBundleFactory(JobInfo jobInfo) throws Exception {
+    return new DockerJobBundleFactory(jobInfo) {
+
+      @Override
+      protected ServerFactory getServerFactory() {
+        return ServerFactory.createWithUrlFactory(
+            (host, port) -> HostAndPort.fromParts("localhost", port).toString());
+        //return InProcessServerFactory.create();
+      }
+
+      @Override
+      protected EnvironmentFactory getEnvironmentFactory(
+          GrpcFnServer<FnApiControlClientPoolService> controlServer,
+          GrpcFnServer<GrpcLoggingService> loggingServer,
+          GrpcFnServer<ArtifactRetrievalService> retrievalServer,
+          GrpcFnServer<StaticGrpcProvisionService> provisioningServiceServer,
+          Source clientSource,
+          IdGenerator idGenerator) {
+        String workerId = "fakeWorkerId";
+        ThreadFactory threadFactory = new ThreadFactoryBuilder().setDaemon(true).build();
+        // Create the SDK harness, and wait until it connects
+        sdkHarnessExecutor = Executors.newSingleThreadExecutor(threadFactory);
+        sdkHarnessExecutorFuture =
+            sdkHarnessExecutor.submit(
+                () -> {
+                  try {
+                    FnHarness.main(
+                        workerId,
+                        PipelineOptionsFactory.create(),
+                        loggingServer.getApiServiceDescriptor(),
+                        controlServer.getApiServiceDescriptor(),
+                        ManagedChannelFactory.createDefault(),
+                        //InProcessManagedChannelFactory.create(),
+                        HarnessStreamObserverFactories.fromOptions(
+                            PipelineOptionsFactory.create()));
+                  } catch (Exception e) {
+                    throw new RuntimeException(e);
+                  }
+                });
+        try {
+          InstructionRequestHandler controlClient =
+              clientSource.take(workerId, Duration.ofSeconds(2));
+          return (container) -> RemoteEnvironment.forHandler(container, controlClient);
+        } catch (Exception ex) {
+          throw new RuntimeException(ex);
+        }
+      }
+    };
+  }
+
+  @Before
+  public void setup() {
+    DockerJobBundleFactory.FACTORY.set(this::createJobBundleFactory);
+  }
+
+  @After
+  public void tearDown() throws Exception {
+    sdkHarnessExecutor.shutdownNow();
+    try {
+      sdkHarnessExecutorFuture.get();
+    } catch (ExecutionException e) {
+      if (e.getCause() instanceof RuntimeException
+          && e.getCause().getCause() instanceof InterruptedException) {
+        // expected
+      } else {
+        throw e;
+      }
+    }
+  }
+
+  private static ArrayList<KV<String, Iterable<Long>>> outputValues = new ArrayList<>();
```

Comment:
nit: you can move this into the test since the anonymous DoFn can refer to it there.

Reply:
That's what I started with, but it needs to be a static variable.